set-client/src/main/java/eu/europeana/set/client/connection/BaseApiConnection.java

```diff
@@ -86,23 +87,83 @@ public void setAuthenticationHandler(AuthenticationHandler auth) {
         this.authHandler = auth;
     }
 
-
     /**
      * Fetches the get user set response (GET request)
      *
+     * If caching object is present, will add the caching headers in the requests
+     *    1. If the ETag is present in the caching set the HTTP request header with If-None-Match.
+     *    2. If the LastModified is present in the caching set the HTTP request header with If-Modified-Since
+     *
+     *    When receiving the response, if the response is a HTTP 200 then a User Set needs to be parsed and
+     *    the ResourceCaching needs to be set with all the caching Response headers.
+     *
+     *    If the response is 304 nothing to be done
+     *    For 404 throw an exception
+     *
      * @param url
      * @return
      * @throws SetApiClientException
      */
-    protected UserSet getUserSetResponse(String url) throws SetApiClientException {
+    protected Optional<UserSet> getUserSetResponse(String url, Optional<ResourceCaching> caching) throws SetApiClientException {
         try {
             LOGGER.trace("Call to Get UserSet API (GET) : {}.", url);
-            return parseSetApiResponse(getHttpConnection().get(url, null, getAuthenticationHandler()), new ArrayList<>(Arrays.asList(HttpStatus.SC_OK, HttpStatus.SC_NOT_MODIFIED)));
+            HttpGet get = getHttpConnection().getHttpRequest(url, null, getAuthenticationHandler());
+            ResourceCaching apiCaching = setCachingHeaders(caching, get);
+            HttpResponseHandler response = getHttpConnection().executeHttpClient(get);
+            String responseBody = response.getResponse();
+            if (response.getStatus() == HttpStatus.SC_OK) {
+                updateResourceCaching(response.getCachingHeaders(), apiCaching);
+                return Optional.of(mapper.readValue(responseBody, UserSet.class));
+            } else if (response.getStatus() == HttpStatus.SC_NOT_MODIFIED) {
+               return Optional.empty();
+            }
+            else {
+                AbstractUserSetApiResponse errorResponse = mapper.readValue(responseBody, AbstractUserSetApiResponse.class);
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(ERROR_MESSAGE + " {} ", errorResponse.getMessage());
+                }
+                throw new SetApiClientException(ERROR_MESSAGE + errorResponse.getMessage(), response.getStatus());
+            }
         } catch (IOException e) {
             throw new SetApiClientException(ERROR_MESSAGE + e.getMessage(), HttpStatus.SC_INTERNAL_SERVER_ERROR, e);
         }
     }
 
+    private ResourceCaching setCachingHeaders(Optional<ResourceCaching> caching, HttpGet get) {
```

Comment:
should rename caching  to cachingOptional

Reply:
done